tempesta_fw/classifier/req_conn_limit.c

```diff
@@ -251,11 +257,109 @@ rcl_req_limit(RclAccount *ra, struct sock *sk)
 }
 
 static int
+rcl_http_uri_len_limit(const TfwHttpReq *req)
+{
+	/* FIXME: tfw_str_len() iterates over chunks to calculate the length.
+	 * This is too slow. The value must be stored in a TfwStr field. */
+	if (rcl_http_uri_len && tfw_str_len(&req->uri) > rcl_http_uri_len) {
+		TFW_DBG("rcl: http_uri_len limit is reached\n");
+		return TFW_BLOCK;
+	}
+
+	return TFW_PASS;
+}
+
+static int
+rcl_http_field_len_limit(const TfwHttpReq *req)
+{
+	const TfwStr *field, *end;
+
+	if (!rcl_http_field_len)
+		return TFW_PASS;
+
+	TFW_HTTP_FOR_EACH_HDR_FIELD(field, end, req) {
+		if (tfw_str_len(field) > rcl_http_field_len) {
+			TFW_DBG("rcl: http_field_len limit is reached\n");
+			return TFW_BLOCK;
+		}
+	}
+
+	return TFW_PASS;
+}
+
+static int
+rcl_http_body_len_limit(const TfwHttpReq *req)
+{
+	if (rcl_http_body_len && tfw_str_len(&req->body) > rcl_http_body_len) {
+		TFW_DBG("rcl: http_body_len limit is reached\n");
+		return TFW_BLOCK;
+	}
+
+	return TFW_PASS;
+}
+
+static int
+rcl_http_len_limit(const TfwHttpReq *req)
+{
+	int r;
+
+	r = rcl_http_uri_len_limit(req);
+	if (r)
+		return r;
+	r = rcl_http_field_len_limit(req);
+	if (r)
+		return r;
+	r = rcl_http_body_len_limit(req);
+	if (r)
+		return r;
+
+	return 0;
+}
+
+static int
+rcl_http_methods_check(const TfwHttpReq *req)
+{
+	unsigned long m = (1 << req->method);
+
+	if (rcl_http_methods_mask && (rcl_http_methods_mask & m))
+		return TFW_PASS;
+
+	TFW_DBG("rcl: forbidden method: %d (%#lx)\n", req->method, m);
+	return TFW_BLOCK;
+}
+
+static int
 rcl_http_req_handler(void *obj, unsigned char *data, size_t len)
 {
+	int r;
 	TfwConnection *c = (TfwConnection *)obj;
+	TfwHttpReq *req = container_of(c->msg, TfwHttpReq, msg);
 
-	return rcl_account_do(c->sess->cli->sock, rcl_req_limit);
+	r = rcl_account_do(c->sess->cli->sock, rcl_req_limit);
+	if (r)
+		return r;
+	r = rcl_http_len_limit(req);
+	if (r)
+		return r;
+
+	return 0;
+}
+
+static int
+rcl_http_chunk_handler(void *obj, unsigned char *data, size_t len)
+{
+	int r;
+	TfwConnection *c = (TfwConnection *)obj;
+	TfwHttpReq *req = container_of(c->msg, TfwHttpReq, msg);
+
+	r = rcl_http_methods_check(req);
+	if (r)
+		return r;
+	r = rcl_http_len_limit(req);
```

Comment:
Why do you call the same checks set for fully parsed request and only small chunk? This is wrong and inefficient.